scale-decode/src/impls/mod.rs

```diff
@@ -0,0 +1,1093 @@
+// ^ Note: We make sure to support as many as parity-scale-codec's impls do.
+
+/// This takes anything that can decode a stream if items and return an iterator over them.
+fn decode_items_using<'a, 'scale, 'info, D: DecodeItemIterator<'scale, 'info>, T>(
+    decoder: &'a mut D,
+) -> impl Iterator<Item = Result<T, Error>> + 'a
+where
+    T: IntoVisitor,
+    Error: From<<T::Visitor as Visitor>::Error>,
+    D: DecodeItemIterator<'scale, 'info>,
+{
+    let mut idx = 0;
+    std::iter::from_fn(move || {
+        let item = decoder
+            .decode_item(T::into_visitor())
+            .map(|res| res.map_err(|e| Error::from(e).at_idx(idx)));
+        idx += 1;
+        item
+    })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::DecodeAsType;
+    use codec::Encode;
+
+    /// Given a type definition, return type ID and registry representing it.
+    fn make_type<T: scale_info::TypeInfo + 'static>() -> (u32, scale_info::PortableRegistry) {
+        let m = scale_info::MetaType::new::<T>();
+        let mut types = scale_info::Registry::new();
+        let id = types.register_type(&m);
+        let portable_registry: scale_info::PortableRegistry = types.into();
+
+        (id.id(), portable_registry)
+    }
+
+    // For most of our tests, we'll assert that whatever type we encode, we can decode back again to the given type.
+    fn assert_encode_decode_to_with<T, A, B>(a: &A, b: &B)
+    where
+        A: Encode,
+        B: DecodeAsType + PartialEq + std::fmt::Debug,
+        T: scale_info::TypeInfo + 'static,
+    {
+        let (type_id, types) = make_type::<T>();
+        let encoded = a.encode();
+        let decoded =
+            B::decode_as_type(&mut &*encoded, type_id, &types).expect("should be able to decode");
+        assert_eq!(&decoded, b);
+    }
+
+    // Normally, the type info we want to use comes along with the type we're encoding.
+    fn assert_encode_decode_to<A, B>(a: &A, b: &B)
+    where
+        A: Encode + scale_info::TypeInfo + 'static,
+        B: DecodeAsType + PartialEq + std::fmt::Debug,
+    {
+        assert_encode_decode_to_with::<A, A, B>(a, b);
+    }
+
+    // Most of the time we'll just make sure that we can encode and decode back to the same type.
+    fn assert_encode_decode_with<T, A>(a: &A)
+    where
+        A: Encode + DecodeAsType + PartialEq + std::fmt::Debug,
+        T: scale_info::TypeInfo + 'static,
+    {
+        assert_encode_decode_to_with::<T, A, A>(a, a)
+    }
+
+    // Most of the time we'll just make sure that we can encode and decode back to the same type.
+    fn assert_encode_decode<A>(a: &A)
+    where
+        A: Encode + scale_info::TypeInfo + 'static + DecodeAsType + PartialEq + std::fmt::Debug,
+    {
+        assert_encode_decode_to(a, a)
+    }
+
+    // Test that something can be encoded and then DecodeAsFields will work to decode it again.
+    fn assert_encode_decode_as_fields<Foo>(foo: Foo)
```

Comment:
nit: Maybe this should also be placed under `#[cfg(feature = "derive")]` since that's the only place we test it?

Reply:
I don't quite follow? I've put the `cfg(feature = "derive")` on any test which requires the `derive` feature is all (ie that has a `#[derive(DecodeAsType]]` macro usage in it). Though interestingly, running `cargo test --no-default-features` still seems to run those tests...

Reply:
Ah; you need to run `cargo test` at the crate level and not workspace level...

And then the feature is disabled anyway. Maybe I'll just remove all of the cfg's and expect all features to be enabled when running tests

Reply:
Ok; removed all of the cfg's around tests :)